src/zjs_callbacks.c

```diff
@@ -387,64 +386,86 @@ void print_callbacks(void)
+void zjs_call_callback(int32_t id, void* data, uint32_t sz)
 {
-    if (cb_map[i]->type == CALLBACK_TYPE_JS) {
-        if (cb_map[i]->js->func_list == NULL && jerry_value_is_function(cb_map[i]->js->js_func)) {
-            uint32_t argc = 0;
-            jerry_value_t ret_val;
-            jerry_value_t* args = NULL;
-
-            if (cb_map[i]->js->pre) {
-                args = cb_map[i]->js->pre(cb_map[i]->js->handle, &argc);
-            }
-
-            DBG_PRINT("calling callback id %ld with %lu args\n", cb_map[i]->js->id, argc);
-            ret_val = jerry_call_function(cb_map[i]->js->js_func, cb_map[i]->js->this, args, argc);
-            if (jerry_value_has_error_flag(ret_val)) {
-                DBG_PRINT("callback %ld returned an error\n", i);
-            }
-            if (cb_map[i]->js->post) {
-                cb_map[i]->js->post(cb_map[i]->js->handle, &ret_val);
-            }
-            if (cb_map[i]->js->once) {
-                zjs_remove_callback(i);
-            }
-        } else {
-            int j;
-            uint32_t argc = 0;
-            jerry_value_t ret_val;
-            jerry_value_t* args = NULL;
-
-            if (cb_map[i]->js->pre) {
-                args = cb_map[i]->js->pre(cb_map[i]->js->handle, &argc);
-            }
-
-            DBG_PRINT("calling callback list id %ld with %lu args\n", cb_map[i]->js->id, argc);
+    if(cb_map[id]) {
+        if (cb_map[id]->type == CALLBACK_TYPE_JS) {
+            if (cb_map[id]->js->func_list == NULL && jerry_value_is_function(cb_map[id]->js->js_func)) {
+                jerry_value_t ret_val;
+                // Single function callback
+                if (sz) {
+                    ret_val = jerry_call_function(cb_map[id]->js->js_func, cb_map[id]->js->this, data, sz);
+                } else {
+                    ret_val = jerry_call_function(cb_map[id]->js->js_func, cb_map[id]->js->this, NULL, 0);
+                }
+                if (cb_map[id]->js->post) {
+                    cb_map[id]->js->post(cb_map[id]->js->handle, &ret_val);
+                }
+                if (cb_map[id]->js->once) {
+                    zjs_remove_callback(id);
+                }
+            } else {
+                // Function list callback
+                int i;
+                jerry_value_t ret_val;
 
-            for (j = 0; j < cb_map[i]->js->num_funcs; ++j) {
-                ret_val = jerry_call_function(cb_map[i]->js->func_list[j], cb_map[i]->js->this, args, argc);
-                if (jerry_value_has_error_flag(ret_val)) {
-                    DBG_PRINT("callback %ld returned an error for function[%i]\n", i, j);
+                for (i = 0; i < cb_map[id]->js->num_funcs; ++i) {
+                    if (sz) {
+                        ret_val = jerry_call_function(cb_map[id]->js->func_list[i], cb_map[id]->js->this, data, sz);
+                    } else {
+                        ret_val = jerry_call_function(cb_map[id]->js->js_func, cb_map[id]->js->this, NULL, 0);
```

Comment:
I think you meant to call func_list[i] here too? And again this should be the same as above so the if/else can be dropped I think.